packages/abi-coder/src/fragments/param-type.ts

```diff
@@ -0,0 +1,175 @@
+import { Logger } from '@ethersproject/logger';
+import { defineReadOnly } from '@ethersproject/properties';
+
+import { arrayRegEx, enumRegEx, structRegEx } from '../abi-coder';
+
+const version = 'abi/5.6.4';
+export const abiLogger = new Logger(version);
+
+export interface JsonFragmentType {
+  readonly name?: string;
+  readonly indexed?: boolean;
+  readonly type: string;
+  readonly components?: ReadonlyArray<JsonFragmentType>;
+  readonly typeArguments?: ReadonlyArray<JsonFragmentType>;
+}
+
+const CONSTRUCTOR_GUARD = {};
+
+function populate(object: ParamType, params: { type?: string } & ParamTypeProps) {
+  Object.keys(params).forEach((key) => {
+    const paramTypeKey = key as keyof ParamTypeProps;
+    const value = params[paramTypeKey];
+
+    defineReadOnly(object, paramTypeKey, value);
+  });
+}
+
+const paramTypeArray = /^(.*)\[([0-9]*)\]$/;
+
+export interface ParamTypeProps {
+  // The local name of the parameter (of null if unbound)
+  readonly name?: string;
+
+  // The fully qualified type (e.g. "address", "tuple(address)", "uint256[3][]"
+  readonly type: string;
+
+  // The base type (e.g. "address", "tuple", "array")
+  readonly baseType?: string;
+
+  // Indexable Paramters ONLY (otherwise null)
+  readonly indexed?: boolean;
+
+  // Tuples ONLY: (otherwise null)
+  //  - sub-components
+  readonly components?: Array<ParamType>;
+
+  // Tuples ONLY: (otherwise null)
+  //  - sub-components
+  readonly typeArguments?: Array<ParamType>;
+
+  // Arrays ONLY: (otherwise null)
+  //  - length of the array (-1 for dynamic length)
+  //  - child type
+  readonly arrayLength?: number;
+  readonly arrayChildren?: ParamType;
+
+  readonly isParamType?: boolean;
+}
+
+export class ParamType implements ParamTypeProps {
+  readonly name?: string;
+  readonly type!: string;
+  readonly baseType?: string;
+  readonly indexed?: boolean;
+  readonly components?: Array<ParamType>;
+  readonly typeArguments?: Array<ParamType>;
+  readonly arrayLength?: number;
+  readonly arrayChildren?: ParamType;
+  readonly isParamType?: boolean;
+
+  constructor(constructorGuard: typeof CONSTRUCTOR_GUARD, params: ParamTypeProps) {
+    if (constructorGuard !== CONSTRUCTOR_GUARD) {
```

Comment:
This is a very used approach on the ETH lib, we normally don't implement it, why do you think is a good choice to add this?

Reply:
This approach is good for safety of classes. As it's "impossible" to have same reference of  `CONSTRUCTOR_GUARD` object outside of this file, creating this class only works using `ParamType.fromObject` method.

This blocks to extend ParamType and try to do something like
```
class HackyParamType extends ParamType {
  constructor(params) {
    super(params);

    // any hacky code
  }
}
```

Combining this with `Object.freeze(this);` in the end of constructor, makes the class `immutable and unextendable`.

Reply:
for this class, I will remove the safety `CONSTRUCTOR_GUARD` part, as it just exports methods signatures. But should consider this approach to other more criticals classes that handle things like prices, bytePrices, gasLimit etc..